### PR TITLE
Better occurrence checking

### PR DESCRIPTION
**Pull request by SquidDev**
See the commit messages for a description. This should allow us to implement some more powerful passes.

### CODE REVIEW

**Comment by zardyh** [comment]
> See the commit messages for a description.

smh. I've done it too, but smh.

File: src/Core/Occurrence.hs [hunk 1]
```diff
@@ -2,126 +2,182 @@
...
-import Control.Arrow
-
-import Syntax (Var(..), Resolved)
 import Pretty
 
-type OccAtom a = C.Atom (OccursVar a)
-type OccTerm a = C.Term (OccursVar a)
-type OccType a = C.Type (OccursVar a)
-type OccStmt a = C.Stmt (OccursVar a)
+data Occurrence = Dead | Once | OnceLambda | Multi | MultiLambda
```
Comment (zardyh) [inline src/Core/Occurrence.hs]: Have you been reading _Secrets of the GHC Inliner_?

File: src/Core/Occurrence.hs [hunk 2]
```diff
@@ -2,126 +2,182 @@
...
+defOcc = MultiLambda
+
+escapeLambda :: Occurrence -> Occurrence
+escapeLambda Once = OnceLambda
+escapeLambda Multi = MultiLambda
+escapeLambda x = x
+
+instance Semigroup Occurrence where
+  x <> Dead = x
+  Dead <> x = x
```
Comment (zardyh) [inline src/Core/Occurrence.hs]: Is this a Monoid with `Dead` as `mempty`?

**Comment by SquidDev** [comment]
> > See the commit messages for a description.
>
> smh. I've done it too, but smh.

Let's be honest, the commit messages aren't that much more useful.

File: src/Core/Occurrence.hs [hunk 3]
```diff
@@ -2,126 +2,182 @@
...
-depends :: IsVar a => a -> OccursMap -> OccursVar a
-depends v ss = OccursVar v (fromMaybe 0 (toVar v `Map.lookup` ss))
-
-convert :: Functor f => f a -> f (OccursVar a)
-convert = fmap (flip OccursVar 1)
+tagOccursVar :: IsVar a => [AnnStmt b a] -> [AnnStmt b (OccursVar a)]
+tagOccursVar = snd . tagOccurStmt const OccursVar
+
+tagOccursMap :: IsVar a => [AnnStmt b a] -> [AnnStmt OccursMap a]
+tagOccursMap = snd . tagOccurStmt (flip const) const
```
Comment (zardyh) [inline src/Core/Occurrence.hs]: Nice expression, even I'm confused.

File: src/Core/Occurrence.hs [hunk 4]
```diff
@@ -2,126 +2,182 @@
...
 
-type OccAtom a = C.Atom (OccursVar a)
-type OccTerm a = C.Term (OccursVar a)
-type OccType a = C.Type (OccursVar a)
-type OccStmt a = C.Stmt (OccursVar a)
+data Occurrence = Dead | Once | OnceLambda | Multi | MultiLambda
+  deriving (Show, Eq, Ord, Data)
+
+defOcc :: Occurrence
+defOcc = MultiLambda
```
Comment (zardyh) [inline src/Core/Occurrence.hs]: See below.

File: src/Core/Occurrence.hs (hunk 1 above)
Comment (SquidDev) [reply src/Core/Occurrence.hs]: Yeah. It doesn't include any of the loop breaker information (yet), but this allows us to reuse it for lifting.

File: src/Core/Occurrence.hs (hunk 4 above)
Comment (zardyh) [reply src/Core/Occurrence.hs]: It ended up being above! I mean the comment about Monoid.

**Comment by zardyh** [comment]
We're a lot better at the complicated things than we are at the simple things.

**Pull request merged**